### ISSUE TITLE
Model uploaded multiple times to GPU memory when training multitask?

### ISSUE THREAD

**Issue by FremyCompany** [UKPLab/sentence-transformers#854]
It seems that when I am training with multiple training objectives, the Transformer model is uploaded one time per objective in the GPU memory, despite being shared by all the losses. This is quickly exhausting such memory as more objectives are being added.

Does that sound like something that is expected given your code?

Is there a way to prevent that?

**Comment by FremyCompany** [comment]
After a more precise investigation, the issue is that the Optimizer instances are not shared across losses. ADAM stores mean/variance per parameter, and this results in all the parameters of the BERT model to have such mean/variance stored in the GPU as many times as there are training objectives in use.

Unfortunately, reusing the same loss does not work because an new optimizer is created anyway per objective, but this can be fixed using a code like this:
```py
        for loss_model in loss_models:

            # re-use the same optimizer if possible
            loss_first_index = loss_models.index(loss_model)
            if loss_first_index < len(optimizers):
                # we already have an optimizer for that loss
                optimizer = optimizers[loss_first_index]
            else:
                # we need a new optimizer for that loss
                param_optimizer = list(loss_model.named_parameters())
                no_decay = ['bias', 'LayerNorm.bias', 'LayerNorm.weight']
                optimizer_grouped_parameters = [
                    {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)], 'weight_decay': weight_decay},
                    {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
                ]
                optimizer = optimizer_class(optimizer_grouped_parameters, **optimizer_params)
            
            scheduler_obj = self._get_scheduler(optimizer, scheduler=scheduler, warmup_steps=warmup_steps, t_total=num_train_steps)

            optimizers.append(optimizer)
            schedulers.append(scheduler_obj)
```

As an additional workaround, I am storing in the loss a counter such that I can switch the parameters I am using depending on the current training objective (since I know there will be one batch per objective, always in the same order).

```py
        # move to the next loss parameters for the next batch
        self.current_step = (self.current_step + 1) % self.step_count
```

This is a bit hacky, a cleaner solution would be better. I am not sure how easy it would be to share the weights of the optimizer instead where possible, instead of sharing the optimizer entirely.

Anyway, food for thought I guess ;-)

**Comment by nreimers** [comment]
Thanks for the interesting observation.

The question is, if there are negative downsides if you share the optimizer for different types of loss?

**Comment by FremyCompany** [comment]
I would assume sharing is better. The goal of Adam is to smooth the gradients. Having two separate Adam running in parallel over the same weights (the model) should be detrimental as they don't take into account the gradient of the other in their running mean, and wrongly estimate how big the learning rate should be. That's my understanding at least.
